### PR TITLE
bug(tianmu):improve the code quality and refactoring derived table optimize for tianmu(#1277, #1276, #1258)

### PR DESCRIPTION
**Pull request by isredstar**
## Summary about this PR

Issue Number: close #1247,#1276,#1258
refactor derived table optimize for tianmu
improve the code quality to modify the magic number
## Tests Check List

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

## Changelog

- [x] New Feature
- [x] Bug Fix
- [ ] Performance Improvement
- [ ] Build/Testing/CI/CD
- [ ] Documentation
- [ ] Not for changelog (changelog entry is not required)

## Documentation

- [x] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features

### CODE REVIEW

File: sql/sql_optimizer.cc [hunk 1]
```diff
@@ -137,371 +137,371 @@ static uint32 get_key_length_tmp_table(Item *item);
...
-  }
+      tables_list= select_lex->get_table_list();
+      
+      /* dump_TABLE_LIST_graph(select_lex, select_lex->leaf_tables); */
+      /*
+        Run optimize phase for all derived tables/views used in this SELECT,
+        including those in semi-joins.
+      */
+      if (select_lex->materialized_derived_table_count)
+        {
```
Comment (lujiashun) [inline sql/sql_optimizer.cc:220]: blank space number was more than expected.

File: sql/sql_optimizer.cc [hunk 2]
```diff
@@ -137,371 +137,371 @@ static uint32 get_key_length_tmp_table(Item *item);
...
-    choose the one that is to be retrieved first and substitute
-    all references to these in where condition for a reference for
-    the selected field.
-  */
-  if (where_cond)
-  {
-    where_cond= substitute_for_best_equal_field(where_cond, cond_equal,
-                                                map2table);
-    if (thd->is_error())
+      // JOIN::optimize_rollup() may set quick_group=0, and we must not undo that.
```
Comment (lujiashun) [inline sql/sql_optimizer.cc:428]: "must undo" or "must not undo"?

File: sql/sql_optimizer.h [hunk 3]
```diff
@@ -55,6 +55,15 @@ typedef struct st_rollup
   List<Item> *fields;
 } ROLLUP;
 
+enum class OptimizePhase // for Tianmu to indicate which optimization phase.
```
Comment (lujiashun) [inline sql/sql_optimizer.h:58]: refer to googel style: https://google.github.io/styleguide/cppguide.html#Enumerator_Names

File: sql/sql_optimizer.cc [hunk 4]
```diff
@@ -10444,14 +10435,14 @@ static bool internal_remove_eq_conds(THD *thd, Item *cond,
     }
     if (cond->const_item())
     {
-      if (part!=1 || cond->type()!=Item::SUBSELECT_ITEM)
-      {//TIANMU UPGRADE
+      if (phase !=OptimizePhase::Before_LOJ_Transform || cond->type()!=Item::SUBSELECT_ITEM)
```
Comment (lujiashun) [inline sql/sql_optimizer.cc]: add a blank space after '!="

**Review by lujiashun — COMMENTED** [review]
suggest to refer to this script to format your code.( you should have installed clang-format, clang-format-diff.py)
```
#/bin/bash
echo $0
cd ..
cur=`pwd`"/"
cd  storage/tianmu
echo $cur
git status .|grep modified |awk '{print $2}'|xargs -i -n 1 clang-format  --style=file -i {}
git diff HEAD -U0 --no-color --src-prefix=$cur --dst-prefix=$cur | clang-format-diff.py -i -style=file
```

File: sql/sql_optimizer.cc (hunk 1 above)
Comment (isredstar) [reply sql/sql_optimizer.cc:220]: ACK

File: sql/sql_optimizer.cc (hunk 4 above)
Comment (isredstar) [reply sql/sql_optimizer.cc]: ACK

File: sql/sql_optimizer.h (hunk 3 above)
Comment (isredstar) [reply sql/sql_optimizer.h:58]: ACK

File: sql/sql_optimizer.cc (hunk 2 above)
Comment (isredstar) [reply sql/sql_optimizer.cc:428]: this is MySQL source code: `must not undo`

File: sql/sql_lex.h [hunk 5]
```diff
@@ -655,10 +655,10 @@ class st_select_lex_unit: public Sql_alloc
   bool prepare(THD *thd, Query_result *result, ulonglong added_options,
                ulonglong removed_options);
   bool optimize(THD *thd);
-  //TIANMU UPGRADE END
+  
   int optimize_for_tianmu();
-  int optimize_after_tianmu();
-  //END
+  int after_optimize_tianmu();
```
Comment (wisehead) [inline sql/sql_lex.h]: why rename the function name? what's the difference?

**Review by wisehead — CHANGES_REQUESTED** [review]

File: sql/sql_lex.h (hunk 5 above)
Comment (isredstar) [reply sql/sql_lex.h]: The `optimize_after_tianmu`  has been as a variable name，so I think it is not suitable as a function name.

Comment (wisehead) [reply sql/sql_lex.h]: a function name should start with a verb. like set_value, update_row, I suggest rename the variable name.

**Review by wisehead — APPROVED** [review]

**Review by lujiashun — APPROVED** [review]
LGTM

**Pull request merged**